File: lib/core/models/live_service.dart
// ──────────────────────────────────────────────────────────────────────────────
// LIVE SERVICE MODEL
// ──────────────────────────────────────────────────────────────────────────────

class LiveService {
  const LiveService({
    required this.id,
    required this.title,
    required this.status,
    required this.viewerCount,
    this.description,
    this.streamUrl,
    this.scheduledAt,
  });

  final String id;
  final String title;
  final String status; // SCHEDULED | LIVE | ENDED
  final int viewerCount;
  final String? description;
  final String? streamUrl;
  final String? scheduledAt;

  bool get isLive => status == 'LIVE';

  factory LiveService.fromJson(Map<String, dynamic> json) {
    return LiveService(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? 'Live Service',
      status: json['status'] as String? ?? 'SCHEDULED',
      viewerCount: json['viewerCount'] as int? ?? 0,
      description: json['description'] as String?,
      streamUrl: json['streamUrl'] as String?,
      scheduledAt: json['scheduledAt'] as String?,
    );
  }
}

class LiveCurrent {
  const LiveCurrent({
    required this.isLive,
    this.current,
    this.upcoming = const [],
  });

  final bool isLive;
  final LiveService? current;
  final List<LiveService> upcoming;

  factory LiveCurrent.fromJson(Map<String, dynamic> json) {
    final currentJson = json['current'] as Map<String, dynamic>?;
    final upcomingJson = (json['upcoming'] as List?)
            ?.map((e) => LiveService.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [];
    return LiveCurrent(
      isLive: json['isLive'] as bool? ?? false,
      current: currentJson != null ? LiveService.fromJson(currentJson) : null,
      upcoming: upcomingJson,
    );
  }
}

File: backend/src/modules/admin/live/live.service.ts
import prisma from '../../../config/database';
import { ApiError } from '../../../utils/apiError';
import { LiveServiceStatus } from '../../../generated/prisma/client';

function cw(churchId: string | null | undefined) { return churchId ? { churchId } : {}; }
function requireChurch(churchId: string | null | undefined): string {
  if (!churchId) throw ApiError.badRequest('Church context required');
  return churchId;
}

export const adminLiveService = {
  async list(churchId: string | null | undefined, opts: { page: number; limit: number; status?: string }) {
    const where: any = { ...cw(churchId) };
    if (opts.status) where.status = opts.status as LiveServiceStatus;
    const skip = (opts.page - 1) * opts.limit;
    const [services, total] = await Promise.all([
      prisma.liveService.findMany({ where, orderBy: { scheduledAt: 'desc' }, skip, take: opts.limit }),
      prisma.liveService.count({ where }),
    ]);
    return { services, total, page: opts.page, limit: opts.limit, totalPages: Math.ceil(total / opts.limit) };
  },

  async create(churchId: string | null | undefined, data: { title: string; description?: string; streamUrl?: string; scheduledAt: string }) {
    const id = requireChurch(churchId);
    return prisma.liveService.create({ data: { ...data, churchId: id, scheduledAt: new Date(data.scheduledAt) } });
  },

  async update(churchId: string | null | undefined, serviceId: string, data: any) {
    const cId = requireChurch(churchId);
    const existing = await prisma.liveService.findFirst({ where: { id: serviceId, churchId: cId } });
    if (!existing) throw ApiError.notFound('Live service not found');
    if (data.scheduledAt) data.scheduledAt = new Date(data.scheduledAt);
    return prisma.liveService.update({ where: { id: serviceId }, data });
  },

  async delete(churchId: string | null | undefined, serviceId: string) {
    const cId = requireChurch(churchId);
    const existing = await prisma.liveService.findFirst({ where: { id: serviceId, churchId: cId } });
    if (!existing) throw ApiError.notFound('Live service not found');
    await prisma.liveService.delete({ where: { id: serviceId } });
  },

  async goLive(churchId: string | null | undefined, serviceId: string) {
    const cId = requireChurch(churchId);
    const existing = await prisma.liveService.findFirst({ where: { id: serviceId, churchId: cId } });
    if (!existing) throw ApiError.notFound('Live service not found');
    return prisma.liveService.update({ where: { id: serviceId }, data: { status: LiveServiceStatus.LIVE, startedAt: new Date() } });
  },

  async endService(churchId: string | null | undefined, serviceId: string) {
    const cId = requireChurch(churchId);
    const existing = await prisma.liveService.findFirst({ where: { id: serviceId, churchId: cId } });
    if (!existing) throw ApiError.notFound('Live service not found');
    return prisma.liveService.update({ where: { id: serviceId }, data: { status: LiveServiceStatus.ENDED, endedAt: new Date() } });
  },
};
